lib/pbench/server/cache_manager.py

```diff
@@ -194,6 +181,82 @@ def make_cache_object(dir_path: Path, path: Path) -> CacheObject:
     )
 
 
+class LockRef:
+    """Keep track of a cache lock passed off to a caller"""
+
+    def __init__(self, lock: Path, exclusive: bool = False, wait: bool = True):
+        """Initialize a lock reference
+
+        The lock file is opened in "w+" mode, which is "write update": unlike
+        "r+", this creates the file if it doesn't already exist, but still
+        allows lock conversion between LOCK_EX and LOCK_SH.
+
+        Args:
+            lock: the path of a lock file
+            exclusive: lock for exclusive access
+            wait: wait for lock
+        """
+        self.lock = lock.open("w+")
+        self.locked = False
+        self.exclusive = exclusive
+        self.unlock = True
+        self.wait = wait
+
+    def acquire(self) -> "LockRef":
+        """Acquire the lock"""
+
+        cmd = fcntl.LOCK_EX if self.exclusive else fcntl.LOCK_SH
+        if not self.wait:
+            cmd |= fcntl.LOCK_NB
+        fcntl.lockf(self.lock, cmd)
+        self.locked = True
+        return self
+
+    def release(self):
+        """Release the lock and close the lock file"""
+
+        exception = None
+        try:
+            fcntl.lockf(self.lock, fcntl.LOCK_UN)
+            self.lock.close()
+            self.locked = False
+            self.exclusive = False
```

Comment:
I don't see any point in trying to support lock/unlock cycles. Logically, yeah, I suppose it'd make a tiny bit more sense to move `exclusive` to `acquire`, especially since I later added the `upgrade` and `downgrade` -- but at the time I didn't think to go back, and I'm not sure it's really worth worrying about now. There's certainly no way I would get it done today.